Feature/json logging

PeterDSteinberg opened the pull request:
This PR adds a json stream handler and a default parameter, --json (-j), to the cli.  See also the comment by @srossross on this PR:https://github.com/Anaconda-Server/anaconda-build/pull/124 on anaconda-build.

The json handler in anaconda-build/anaconda-client related changes involve eventually merging:
- [anaconda-build PR 124](https://github.com/Anaconda-Server/anaconda-build/pull/124)
- [anaconda-client PR 226](https://github.com/Anaconda-Server/anaconda-client/pull/226)
- This clyent PR

Current outstanding issues are (TODO):
- Exceptions are not being stringified properly into jsons.  It appears they are silenced.
- The current set up requires different calling arguments to log.info depending on whether the handler is a json stream handler or not.  The difficulty is that the non-json stream handlers expect the message string to be formatted with args by the percent (%) style of string formatting, while we hope to do the new style of curly brace string formatting.

Commit:
new json stream handler

Commit:
changes to json logging

srossross commented on clyent/__init__.py:
```diff
@@ -96,7 +96,9 @@ def add_default_arguments(parser, version=None):
                         const='never',
                         help='never display with colors')
-
+    ogroup.add_argument('-j', '--json',
```
I don't think we need a short option here. This should only be used as a programmatic interface.

Commit:
remove default --json arg; move to logging adapter for json

PeterDSteinberg commented:
I removed the --json argument from default arguments because there is apparently some conflict over the meaning of --json, whether it jsonifies the output logging or the help commands.  With --json in default arguments in clyent, when running ```anaconda worker run --json``` I was getting a jsonified help rather than jsonified logging output.  I think it makes sense to just put the --json argument in more specialized modules than in default args.

Also in the latest commit, I moved to a logging adapter instead of handler for the json stream.

srossross commented:
I am getting this error:

```
(anaconda.org)mr_chomps:clyent sean$ anaconda --json whoami 
Traceback (most recent call last):
  File "/Users/sean/anaconda/envs/anaconda.org/bin/anaconda", line 9, in <module>
    load_entry_point('anaconda-client==1.0.2', 'console_scripts', 'anaconda')()
  File "/Users/sean/anaconda/envs/anaconda.org/lib/python2.7/site-packages/pkg_resources.py", line 343, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/Users/sean/anaconda/envs/anaconda.org/lib/python2.7/site-packages/pkg_resources.py", line 2354, in load_entry_point
    return ep.load()
  File "/Users/sean/anaconda/envs/anaconda.org/lib/python2.7/site-packages/pkg_resources.py", line 2060, in load
    entry = __import__(self.module_name, globals(),globals(), ['__name__'])
  File "/Users/sean/anaconda/envs/anaconda.org/lib/python2.7/site-packages/binstar_client/scripts/cli.py", line 19, in <module>
    from clyent.logs import setup_logging
  File "/Users/sean/Documents/workspace/clyent/clyent/logs/__init__.py", line 12, in <module>
    from .handlers import ColorStreamHandler, JsonStreamAdapter
  File "/Users/sean/Documents/workspace/clyent/clyent/logs/handlers.py", line 61, in <module>
    class JsonStreamAdapter(logging.Adapter):
AttributeError: 'module' object has no attribute 'Adapter'

```

srossross commented on clyent/logs/__init__.py:
```diff
@@ -39,6 +39,8 @@ def setup_logging(logger, level, use_color=None, logfile=None, show_tb=False):
 
     exceptions = (errors.ClyentError, KeyboardInterrupt)
     shndlr = ColorStreamHandler(show_tb=show_tb, exceptions=exceptions)
+    if as_json:
```
I think this would be better suited as a `Formatter` class.

PeterDSteinberg commented:
I'll work on this some more.  That was a confusing commit.

Commit:
use logging.handler in place of logging.loggingAdapter

srossross commented:
> I removed the --json argument from default arguments because there is apparently 
> some conflict over the meaning of --json

@PeterDSteinberg I think it would be useful here to still add a `--json-output` argument to the default arguments. that way all commands can have JSON output.

PeterDSteinberg commented:
sounds good @srossross  (the --json-output flag in default arguments).  I'll make that change.

srossross commented on clyent/logs/handlers.py:
```diff
@@ -54,6 +58,21 @@ def emit(self, record):
             print(msg, file=stream)
 
 
+class JsonStreamHandler(logging.Handler):
```
Is there any reason not to use a `formatter` class here?

https://docs.python.org/2/library/logging.html#formatter-objects

srossross replied:
(Not a big deal, just asking)

srossross commented on clyent/logs/__init__.py:
```diff
@@ -38,9 +38,14 @@ def setup_logging(logger, level, use_color=None, logfile=None, show_tb=False):
+        shndlr = JsonStreamHandler()
     shndlr.setLevel(level)
     logger.addHandler(shndlr)
+    #if as_json:
```
remove comments.

Commit:
--json-output option for all commands

PeterDSteinberg commented:
I will look into a Formatter class with format method.  This was the first time I had built something custom with logging and it was a bit of a puzzle.

Now the --json-output flag can be given to anaconda on any command, e.g.

```
anaconda --json-output worker register psteinberg/abc
```
```
anaconda --json-output worker run <worker-id>
```

Remaining TODO items:
 * Exceptions are being silenced.  Make sure they are passed through to json.

Commit:
use logging.Formatter instead of Handler; jsonify exceptions

PeterDSteinberg commented:
This latest commit address the TODO issue above.  Now exceptions are stringified into a dictionary that has an 'error' key with a value that is the exception repr and traceback.

srossross commented:
I get this error:
```
Traceback (most recent call last):
  File "/Users/sean/anaconda/envs/anaconda.org/bin/anaconda", line 9, in <module>
    load_entry_point('anaconda-client==1.0.1-55.gd24f685.dirty', 'console_scripts', 'anaconda')()
  File "/Users/sean/Documents/workspace/binstar/binstar_client/scripts/cli.py", line 93, in main
    description=__doc__, version=version)
  File "/Users/sean/Documents/workspace/binstar/binstar_client/scripts/cli.py", line 75, in binstar_main
    return args.main(args)
  File "/Users/sean/Documents/workspace/binstar/binstar_client/commands/whoami.py", line 22, in main
    pprint_user(user)
  File "/Users/sean/Documents/workspace/binstar/binstar_client/utils/pprint.py", line 74, in pprint_user
    log.info('Username: %s', user.pop('login'), metadata=1)
  File "/Users/sean/anaconda/envs/anaconda.org/lib/python2.7/logging/__init__.py", line 1165, in info
    self._log(INFO, msg, args, **kwargs)
TypeError: _log() got an unexpected keyword argument 'metadata'
```

PeterDSteinberg commented:
@srossross  I have used the default way of passing extra info to log.info, warn, etc:

```
log.info('Message string {0}', replacement, extra={'metadata': {'a': 1, 'b': 2}})
```
Passing a dict-like as ```extra``` populates the LogRecord dict for each log entry in formatters/handlers.

I could partial the log, debug, warn, etc functions so that we can give metadata instead of extra, i.e. make wrapper functions for each of those functions so we can do:
```
log.info('Message string {0}, replacement, metadata={'a':1,'b':2)
```

stephenakearns commented:
discussed - closing for the time-being
